torch/distributed/pipelining/stage.py

```diff
@@ -1256,33 +1264,46 @@ def __init__(
         stage_index: int,
         num_stages: int,
         device: torch.device,
-        input_args: Union[torch.Tensor, Tuple[torch.Tensor, ...]],
+        input_args: Optional[Union[torch.Tensor, Tuple[torch.Tensor, ...]]] = None,
         output_args: Optional[Union[torch.Tensor, Tuple[torch.Tensor, ...]]] = None,
         group: Optional[dist.ProcessGroup] = None,
         dw_builder: Optional[Callable[[], Callable[..., None]]] = None,
     ):
         super().__init__(submodule, stage_index, num_stages, device, group, dw_builder)
         self.inputs: Optional[List[torch.Tensor]] = None
-
+        self.inputs_meta: Optional[Tuple[torch.Tensor, ...]] = None
         # Note: inputs and submod should ideally be on meta device. We decided not to assert this (yet) becuase it
         # might be breaking for existing users.
-        self.inputs_meta = (
-            (input_args,) if isinstance(input_args, torch.Tensor) else input_args
-        )
-        if output_args is None:
-            try:
-                output_args = submodule(*self.inputs_meta)
-                output_args = tree_map_only(
-                    torch.Tensor, lambda x: x.to("meta"), output_args
-                )
-            except Exception as e:
-                raise RuntimeError(
-                    "Failed to perform pipeline shape inference- are your inputs on the same device as your module?"
-                ) from e
-        assert output_args is not None  # for mypy
-        self._configure_outputs_meta(
-            (output_args,) if isinstance(output_args, torch.Tensor) else output_args
-        )
+        if input_args is None:
+            assert output_args is None, (
+                "If specifying output_args, input_args must also be specified. "
+                "Otherwise, shape inference will be performed at runtime"
```

Comment:
If the "otherwise" a good thing or a bad thing?

Reply:
its a good thing. is it confusing?